docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,292 @@
+  command []String
+  inputs []File
+  syncables []Syncables
+}
+
+type Direct struct {
+  syncables []Syncables
+}
+
+type Syncables struct {
+  src String
+  dest String
+}
+```
+
+#### Jib - Skaffold Sync interface
+
+How a tool like Jib might surface the necessary information to Skaffold
+
+I would expect to add a task like `_jibSkaffoldSyncMap` that will produce
+json output for the skaffold-jib intergration to consume and forward to the sync
+system. And example output might look like:
+
+```
+BEGIN JIB JSON: SYNCMAP/1
+{
+  "generated": [
+    {
+      src: “target/classes”
+      dest: "app/classes",
+    },
+    {
+      src: “target/resources"
+      dest: "app/resources",
+    }
+  ]
+  "direct": [
+    {
+      src: "src/main/extra1",
+      dest: "/"
+    },
+    {
+      src: "src/main/extra2",
+      dest: "/"
+    },
+    {
+      src: ".m2/some/dep/my-dep-SNAPSHOT.jar",
+      dest: "app/dependencies/my-dep-SNAPSHOT.jar"
+    }
+  }
+}
+```
+
+Files in the `generated` section will trigger a partial rebuild of the container 
+(everything before containerization) while files in the `direct` section can
+just be synchronized to the running container without a rebuild of anything.
+
+##### Sync or Rebuild?
+
+Each builder implementing an `auto` builder should be able to decide when a sync
+should be skipped and a rebuild done instead. In the jib case for instance, a
+rebuild will be triggered if:
+- a build file has changed (`build.gradle`, `pom.xml`, etc)
+- a file is deleted
+
+#### Open Issues/Questions
+
+**What about files that have specific permissions?**
+
+Jib allows users to customize file permissions, for instance a file on the
+container can be configured to be 0x456 or something. One option instead of
+dealing with this, is to just make all sync'd files 777? Or we allow passthrough
+of permissions from the build system to the sync system.
+
+**Should we allow the user to configure the auto block?**
+
+Perhaps the user knows something that jib doesnt, and wants to ignore some files
+from synchronization. They might want to do:
+
+```
+sync:
+  auto:
+    ignored:
+    - "src/main/jib/myExtraFilesThatBreakADeployment"
+```
+
+TODO (comment here if you need this be elaborated more on)
+
+## Implementation plan
+
+- [`schemas/<version>.go`] Add `Auto` to the schema under `Sync`
+- Before first build, initialize the `sync` state for builders in `auto` mode, this
+sync state is saved in the builder's specific implementation of `auto`
+
+- On file change in dev mode: 
+```
+if (files were deleted)
+  return REBUILD
+
+if (changes were made to build def)
+  return REBUILD
+
+lastSyncState = syncStates["this project"]
+
+if (if all files changes are in lastSyncState.direct)
+  return SYNC{list of direct files}
+
+newSyncState = buildAndCalculateNewSyncState("this project")
+syncStates["this project"] = newSyncState
+
+diff = diff(lastSyncState, newSyncState)
+return SYNC{files in diff}
+```
+
+## Integration test plan
+
+*TBD*
```

Comment:
```suggestion

This can be a Kind based test, we don't need GCP secrets. 
Hence: 
- need to install gradle or maven on travis integration test jobs outside the cluster for testing the build script
- the test would do the following steps: 
    - spin up a simple java app built by jib with skaffold dev + autosync enabled
    - get pod name
    - change a .java file (should trigger generation of a class file) 
    - check with kubectl inside the pod (if pod not found as rebuild was triggered, fail) that the class file is copied over (compare content with external version) 
```